File: 2015/advent22/advent.py
class Spell:
    GET = []

    def __init__(self, name, mana_cost, timer, damage=0, add_armor=0, add_mana=0, healing=0):
        self.name = name
        self.mana_cost = mana_cost
        self.timer = timer
        self.damage = damage
        self.add_armor = add_armor
        self.healing = healing
        self.add_mana = add_mana
        self.GET.append(self)

    @staticmethod
    def reset():
        for spell in Spell.GET:
            spell.current_timer = 0

class Wizard:
    def __init__(self, hit, mana):
        self.name = "wizard"
        self.hit = hit
        self.mana = mana

    def reset(self):
        self.current_hit = self.hit
        self.current_mana = self.mana
        self.current_armor = 0

class Boss:
    def __init__(self, hit, damage):
        self.name = "boss"
        self.hit = hit
        self.damage = damage

    def reset(self):
        self.current_hit = self.hit

class Game:
    def __init__(self, wizard, boss):
        self.wizard = wizard
        self.boss = boss

    def reset(self):
        Spell.reset()
        self.wizard.reset()
        self.boss.reset()
        self.current_spells = []
        self.current_turn = 0
        self.winner = ""

    def manage_spells(self, turn_wizard):
        damage_boss = 0
        for spell in self.current_spells[:]:
            if spell.current_timer == 0:
                self.wizard.current_armor += spell.add_armor
            self.wizard.current_mana += spell.add_mana
            self.wizard.hit += spell.healing
            damage_boss += spell.damage
            spell.current_timer += 1
            if spell.current_timer == spell.timer:
                spell.current_timer = 0
                self.current_spells.remove(spell)
                self.wizard.current_armor -= spell.add_armor
        self.boss.current_hit -= damage_boss

    def turn_wizard(self, spell):
        self.current_turn += 1
        if spell.current_timer > 0 and spell.current_timer < spell.timer:
            raise Exception("Impossibile usare lo stesso spell più volte contemporaneamente")
        self.wizard.current_mana -= spell.mana_cost
        if self.wizard.current_mana < 0:
            raise Exception("Mana sottozero")
        self.manage_spells(turn_wizard=True)
        if spell.timer == 1:
            self.boss.current_hit -= spell.damage
            self.wizard.current_hit += spell.healing
        else:
            self.current_spells.append(spell)
        if self.boss.current_hit <= 0:
            self.winner = "wizard"

    def turn_boss(self):
        self.current_turn += 1
        self.manage_spells(turn_wizard=False)
        if self.boss.current_hit <= 0:
            self.winner = "wizard"
        else:
            self.wizard.current_hit -= max(1, self.boss.damage - self.wizard.current_armor)
            if self.wizard.current_hit <= 0:
                self.winner = "boss"

MAGIC_MISSILE = Spell("magic missile", mana_cost=53, timer=1, damage=4)
DRAIN = Spell("drain", mana_cost=73, timer=1, damage=2, healing=2)
SHIELD = Spell("shield", mana_cost=113, timer=6, add_armor=7)
POISON = Spell("poison", mana_cost=173, timer=6, damage=3)
RECHARGE = Spell("recharge", mana_cost=229, timer=5, add_mana=101)

def test_game1():
    wizard = Wizard(10, 250)
    boss = Boss(13, 8)
    game = Game(wizard, boss)
    game.reset()
    game.turn_wizard(POISON)
    if wizard.current_hit != 10:
        print("Dopo un turno il giocatore deve avere 10 hit point")
    if wizard.current_mana != 77:
        print("Dopo un turno il giocatore deve avere 77 di mana")
    if boss.current_hit != 13:
        print("Dopo un turno il boss deve avere 10 hit point")
    game.turn_boss()
    if boss.current_hit != 10:
        print("Dopo 2 turni il boss deve avere 10 hit point")
    game.turn_wizard(MAGIC_MISSILE)
    if wizard.current_hit != 2:
        print("Dopo 3 turni il giocatore deve avere 2 hit point")
    if wizard.current_mana != 24:
        print("Dopo 3 turni il giocatore deve avere 24 di mana")
    if boss.current_hit != 3:
        print("Dopo 3 turni il boss deve avere 3 hit point")
    game.turn_boss()
    if game.winner != "wizard":
        print("Dopo 4 turni deve vincere il wizard")
    if boss.current_hit != 0:
        print("Dopo 4 turni il boss deve avere 0 hit point")

def test_game2():
    wizard = Wizard(10, 250)
    boss = Boss(14, 8)
    game = Game(wizard, boss)
    game.reset()
    game.turn_wizard(RECHARGE)
    if wizard.current_hit != 10:
        print("Dopo un turno il giocatore deve avere 10 hit point")
    if wizard.current_mana != 21:
        print("Dopo un turno il giocatore deve avere 21 di mana")
    if boss.current_hit != 14:
        print("Dopo un turno il boss deve avere 14 hit point")
    game.turn_boss()
    if boss.current_hit != 14:
        print("Dopo 2 turni il boss deve avere 14 hit point")
    if wizard.current_mana != 122:
        print("Dopo 2 turni il giocatore deve avere 122 di mana")
    game.turn_wizard(SHIELD)
    if wizard.current_hit != 2:
        print("Dopo 3 turni il giocatore deve avere 2 hit point")
    if wizard.current_mana != 110:
        print("Dopo 3 turni il giocatore deve avere 24 di mana")
    if boss.current_hit != 14:
        print("Dopo 3 turni il boss deve avere 14 hit point")
    game.turn_boss()
    if wizard.current_hit != 1:
        print("Dopo 4 turni il giocatore deve avere 1 hit point")
    if wizard.current_mana != 211:
        print("Dopo 4 turni il giocatore deve avere 211 di mana")
    game.turn_wizard(DRAIN)
    if wizard.current_mana != 239:
        print("Dopo 5 turni il giocatore deve avere 239 di mana")
    if wizard.current_hit != 3:
        print("Dopo 5 turni il giocatore deve avere 3 hit point")
    game.turn_boss()
    if wizard.current_hit != 2:
        print("Dopo 6 turni il giocatore deve avere 2 hit point")
    if wizard.current_mana != 340:
        print("Dopo 6 turni il giocatore deve avere 340 di mana")
    game.turn_wizard(POISON)
    if wizard.current_hit != 2:
        print("Dopo 7 turni il giocatore deve avere 2 hit point")
    if wizard.current_mana != 167:
        print("Dopo 7 turni il giocatore deve avere 167 di mana")
    game.turn_boss()
    if wizard.current_hit != 1:
        print("Dopo 8 turni il giocatore deve avere 1 hit point")
    if wizard.current_mana != 167:
        print("Dopo 8 turni il giocatore deve avere 167 di mana")
    game.turn_wizard(MAGIC_MISSILE)
    if game.winner != "":
        print("Non ci deve essere un vincitore")
    game.turn_boss()
    if game.winner != "wizard":
        print("Dopo tot turni deve vincere il wizard")
    if wizard.current_hit != 1:
        print("Dopo tot turni il giocatore deve avere 1 hit point")
    if wizard.current_mana != 114:
        print("Dopo tot turni il giocatore deve avere 114 di mana")
    if boss.current_hit != -1:
        print("Dopo tot turni il boss deve avere -1 hit point")

def test():
    test_game1()
    test_game2()

if __name__ == "__main__":
    import sys
    if "-t" in sys.argv:
        test()
    else:
        # Questo va (226 poison + magic missile)
        wizard = Wizard(10, 250)
        boss = Boss(13, 8)

        # Questo va 641 magic missile ['recharge', 'shield', 'drain', 'poison']
        wizard = Wizard(10, 250)
        boss = Boss(14, 8)

        wizard = Wizard(50, 500)
        boss = Boss(58, 9)

        game = Game(wizard, boss)
        min_mana = 99999999

        def manage(game, spell, others):
            global min_mana
            if wizard.current_mana < spell.mana_cost:
                return "c"
            game.turn_wizard(spell)
            r = ""
            if game.winner:
                r = "c"
            else:
                game.turn_boss()
                if game.winner:
                    r = "c"
            if r == "c":
                if game.winner == "wizard":
                    mana_spent = spell.mana_cost + sum(s.mana_cost for s in others)
                    if mana_spent < min_mana:
                        min_mana = mana_spent
                    print(min_mana, spell.name, [s.name for s in others])
                game.reset()
                for s in others:
                    x = manage(game, s, [])
                    assert x == ""
            return r

        # 1269 ['P', 'R', 'D', 'P', 'R', 'S', 'P', 'M', 'M', 'M', 'M', 'M', 'M', 'M', 'M', 'M', 'M', 'M', 'M']

        for s0 in Spell.GET:
            print(s0.name)
            game.reset()
            if manage(game, s0, []):
                break
            print("t0", s0.name, wizard.current_mana)
            for s1 in Spell.GET:
                print("  t1", s0.name, s1.name, wizard.current_mana)
                print("  ", s1.name)
                if s1.current_timer > 0:
                    continue
                r = manage(game, s1, [s0])
                if r == "c":
                    continue
                if r == "b":
                    break
                for s2 in Spell.GET:
                    print("    ", s2.name)
                    if s2.current_timer > 0:
                        continue
                    r = manage(game, s2, [s0, s1])
                    if r == "c":
                        continue
                    if r == "b":
                        break
                    for s3 in Spell.GET:
                        print("      ", s3.name)
                        if s3.current_timer > 0:
                            continue
                        r = manage(game, s3, [s0, s1, s2])
                        if r == "c":
                            continue
                        if r == "b":
                            break
                        for s4 in Spell.GET:
                            print("        ", s4.name)
                            if s4.current_timer > 0:
                                continue
                            r = manage(game, s4, [s0, s1, s2, s3])
                            if r == "c":
                                continue
                            if r == "b":
                                break
                            for s5 in Spell.GET:
                                print("          ", s4.name)
                                if s5.current_timer > 0:
                                    continue
                                r = manage(game, s5, [s0, s1, s2, s3, s4])
                                if r == "c":
                                    continue
                                if r == "b":
                                    break
                                for s6 in Spell.GET:
                                    if s6.current_timer > 0:
                                        continue
                                    r = manage(game, s6, [s0, s1, s2, s3, s4, s5])
                                    if r == "c":
                                        continue
                                    if r == "b":
                                        break
                                    for s7 in Spell.GET:
                                        if s7.current_timer > 0:
                                            continue
                                        r = manage(game, s7, [s0, s1, s2, s3, s4, s5, s6])
                                        if r == "c":
                                            continue
                                        if r == "b":
                                            break
                                        for s8 in Spell.GET:
                                            if s8.current_timer > 0:
                                                continue
                                            r = manage(game, s8, [s0, s1, s2, s3, s4, s5, s6, s7])
                                            if r == "c":
                                                continue
                                            if r == "b":
                                                break
                                            for s9 in Spell.GET:
                                                if s9.current_timer > 0:
                                                    continue
                                                r = manage(game, s9, [s0, s1, s2, s3, s4, s5, s6, s7, s8])
                                                if r == "c":
                                                    continue
                                                if r == "b":
                                                    break
                                                for s10 in Spell.GET:
                                                    if s10.current_timer > 0:
                                                        continue
                                                    r = manage(game, s10, [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9])
                                                    if r == "c":
                                                        continue
                                                    if r == "b":
                                                        break
                                                    for s11 in Spell.GET:
                                                        if s11.current_timer > 0:
                                                            continue
                                                        r = manage(game, s11, [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10])
                                                        if r == "c":
                                                            continue
                                                        if r == "b":
                                                            break
                                                        for s12 in Spell.GET:
                                                            if s12.current_timer > 0:
                                                                continue
                                                            r = manage(game, s12, [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11])
                                                            if r == "c":
                                                                continue
                                                            if r == "b":
                                                                break
                                                            for s13 in Spell.GET:
                                                                if s13.current_timer > 0:
                                                                    continue
                                                                r = manage(game, s13, [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12])
                                                                if r == "c":
                                                                    continue
                                                                if r == "b":
                                                                    break
                                                                for s14 in Spell.GET:
                                                                    if s14.current_timer > 0:
                                                                        continue
                                                                    r = manage(game, s14, [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13])
                                                                    if r == "c":
                                                                        continue
                                                                    if r == "b":
                                                                        break
                                                                    for s15 in Spell.GET:
                                                                        if s15.current_timer > 0:
                                                                            continue
                                                                        r = manage(game, s15, [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14])
                                                                        if r == "c":
                                                                            continue
                                                                        if r == "b":
                                                                            break
                                                                        for s16 in Spell.GET:
                                                                            if s16.current_timer > 0:
                                                                                continue
                                                                            r = manage(game, s16, [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15])
                                                                            if r == "c":
                                                                                continue
                                                                            if r == "b":
                                                                                break
                                                                            for s17 in Spell.GET:
                                                                                if s17.current_timer > 0:
                                                                                    continue
                                                                                r = manage(game, s17, [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15, s16])
                                                                                if r == "c":
                                                                                    continue
                                                                                if r == "b":
                                                                                    break
                                                                                for s18 in Spell.GET:
                                                                                    if s18.current_timer > 0:
                                                                                        continue
                                                                                    r = manage(game, s18, [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15, s16, s17])
                                                                                    if r == "c":
                                                                                        continue
                                                                                    if r == "b":
                                                                                        break
                                                                                    for s19 in Spell.GET:
                                                                                        if s19.current_timer > 0:
                                                                                            continue
                                                                                        r = manage(game, s19, [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15, s16, s17, s18])
                                                                                        if r == "c":
                                                                                            continue
                                                                                        if r == "b":
                                                                                            break

File: 2015/advent22/advent.go
package main

import (
	"fmt"
	"math/rand"
)

type Wizard struct {
	hit           int
	mana          int
	armor         int
	current_armor int
	current_hit   int
	current_mana  int
}

func (wizard *Wizard) Reset() {
	wizard.current_hit = wizard.hit
	wizard.current_mana = wizard.mana
	wizard.current_armor = wizard.armor
}

func NewWizard(hit, mana int) *Wizard {
	return &Wizard{hit, mana, 0, 0, 0, 0}
}

type Boss struct {
	hit         int
	damage      int
	current_hit int
}

func (boss *Boss) Reset() {
	boss.current_hit = boss.hit
}

func NewBoss(hit, damage int) *Boss {
	return &Boss{hit, damage, 0}
}

type Game struct {
	wizard         *Wizard
	boss           *Boss
	current_turn   int
	current_spells []*Spell
	winner         string
}

func (game *Game) Reset() {
	game.wizard.Reset()
	game.boss.Reset()
	ResetSpells()
	game.current_spells = []*Spell{}
	game.current_turn = 0
	game.winner = ""
}

func (game *Game) TurnWizard(new_spell *Spell) {
	if new_spell.current_timer > 0 {
		panic("Impossibile usare lo stesso spell più volte nello stesso tempo")
	}
	//fmt.Println("======================")
	//fmt.Println("Spells")
	for n := 0; n < len(game.current_spells); n++ {
		//fmt.Println("  ", game.current_spells[n].name, game.current_spells[n].current_timer)
	}
	//fmt.Println("Casting ", new_spell.name)
	//fmt.Println("Shield.current_timer", GetSpell("Shield").current_timer)
	game.current_turn += 1
	damage_to_boss := 0
	// Spells
	for n := len(game.current_spells) - 1; n >= 0; n-- {
		//spell :=
		//fmt.Println(game.current_spells[n].name, game.current_spells[n].current_timer)
		damage_to_boss += game.current_spells[n].damage
		game.wizard.current_mana += game.current_spells[n].add_mana
		game.current_spells[n].current_timer -= 1
		if game.current_spells[n].current_timer == 0 {
			//fmt.Println("armor_to_remove", game.current_spells[n].add_armor, game.current_spells[n].name)
			game.wizard.current_armor -= game.current_spells[n].add_armor
			game.current_spells = append(game.current_spells[:n], game.current_spells[n+1:]...)
		}
	}
	// Current spell
	if new_spell.timer == 1 {
		game.wizard.current_hit += new_spell.add_hit
		damage_to_boss += new_spell.instant_damage
	} else {
		game.current_spells = append(game.current_spells, new_spell)
		new_spell.current_timer = new_spell.timer
		game.wizard.current_armor += new_spell.add_armor
		//fmt.Println("armor_to_add", new_spell.add_armor)
	}
	//fmt.Println("Spells after")
	for n := 0; n < len(game.current_spells); n++ {
		//fmt.Println("  ", game.current_spells[n].name, game.current_spells[n].current_timer)
	}
	game.wizard.current_mana -= new_spell.mana_cost
	game.boss.current_hit -= damage_to_boss
	if game.wizard.current_mana < 0 {
		game.winner = "boss"
	} else if game.boss.current_hit <= 0 {
		game.winner = "wizard"
	}
	//fmt.Println("Turno", game.current_turn)
	//fmt.Println(game.wizard)
	//fmt.Println(game.boss)
}

func (game *Game) TurnBoss() {
	game.current_turn += 1
	damage_to_boss := 0
	damage_to_wizard := game.boss.damage - game.wizard.armor
	armor_to_remove := 0
	for n := len(game.current_spells) - 1; n >= 0; n-- {
		//fmt.Println(game.current_spells[n].name, game.current_spells[n].current_timer)
		damage_to_boss += game.current_spells[n].damage
		game.wizard.current_hit += game.current_spells[n].add_hit
		game.wizard.current_mana += game.current_spells[n].add_mana
		game.current_spells[n].current_timer -= 1
		if game.current_spells[n].current_timer == 0 {
			armor_to_remove = game.current_spells[n].add_armor
			//fmt.Println("armor_to_remove", armor_to_remove, game.current_spells[n].name)
			game.current_spells = append(game.current_spells[:n], game.current_spells[n+1:]...)
		}
	}
	//fmt.Println("Spells after boss")
	for n := 0; n < len(game.current_spells); n++ {
		//fmt.Println("  ", game.current_spells[n].name, game.current_spells[n].current_timer)
	}
	//fmt.Println("Spells after boss")
	damage_to_wizard -= game.wizard.current_armor
	game.wizard.current_armor -= armor_to_remove
	if damage_to_wizard < 1 {
		damage_to_wizard = 1
	}
	//fmt.Println("Turno", game.current_turn, damage_to_wizard, game.wizard.current_armor)
	//fmt.Println("to boss", damage_to_boss)
	game.boss.current_hit -= damage_to_boss
	//fmt.Println(game.boss)
	if game.boss.current_hit > 0 {
		game.wizard.current_hit -= damage_to_wizard
		if game.wizard.current_hit <= 0 {
			game.winner = "boss"
		}
	} else {
		game.winner = "wizard"
	}
	//fmt.Println("Turno", game.current_turn)
	//fmt.Println(game.wizard)
	//fmt.Println(game.boss)
}

func NewGame(wizard *Wizard, boss *Boss) *Game {
	return &Game{wizard, boss, 0, []*Spell{}, ""}
}

type Spell struct {
	name           string
	mana_cost      int
	damage         int
	add_armor      int
	timer          int
	add_hit        int
	add_mana       int
	current_timer  int
	instant_damage int
}

func (spell *Spell) Reset() {
	spell.current_timer = 0
}

var spells = []*Spell{
	&Spell{"Magic Missile", 53, 0, 0, 1, 0, 0, 0, 4},
	&Spell{"Drain", 73, 0, 0, 1, 2, 0, 0, 2},
	&Spell{"Shield", 113, 0, 7, 6, 0, 0, 0, 0},
	&Spell{"Poison", 173, 3, 0, 6, 0, 0, 0, 0},
	&Spell{"Recharge", 229, 0, 0, 5, 0, 101, 0, 0},
}

func GetSpell(name string) *Spell {
	for n := 0; n < len(spells); n++ {
		if spells[n].name == name {
			return spells[n]
		}
	}
	panic("Spell not found")
}

func ResetSpells() {
	for n := 0; n < len(spells); n++ {
		spells[n].Reset()
	}
}

func main() {
	wizard := NewWizard(50, 500)
	boss := NewBoss(58, 9)
	game := NewGame(wizard, boss)
	min_mana := 99999999
	for {
		mana_spent := 0
		spells_used := []*Spell{}
		game.Reset()
		for game.winner == "" {
			spell := spells[0]
			for {
				spell = spells[rand.Intn(len(spells))]
				if spell.current_timer == 0 {
					break
				}
			}
			spells_used = append(spells_used, spell)
			mana_spent += spell.mana_cost
			game.TurnWizard(spell)
			if game.winner != "" {
				break
			}
			game.TurnBoss()
		}
		if game.winner == "wizard" && mana_spent < min_mana {
			min_mana = mana_spent
			fmt.Println("===========")
			fmt.Println(min_mana)
			mana := 0
			for n := 0; n < len(spells_used); n++ {
				fmt.Println(spells_used[n])
				mana += spells_used[n].mana_cost
			}
			fmt.Println(mana)
		}
	}
	fmt.Println(min_mana)
}
